## Part 1: Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [2]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [3]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[['Attrition', 'Department']]
y_df

Attrition              Department
0          Yes                   Sales
1           No  Research & Development
2          Yes  Research & Development
3           No  Research & Development
4           No  Research & Development
...        ...                     ...
1465        No  Research & Development
1466        No  Research & Development
1467        No  Research & Development
1468        No                   Sales
1469        No  Research & Development

[1470 rows x 2 columns]

In [18]:
# Print the unique values from the "Department" column in y_df dataframe
print(y_df['Department'].unique())

['Sales' 'Research & Development' 'Human Resources']


In [4]:
# Create a list of at least 10 column names to use as X data
X_list = ['Age', 'DistanceFromHome', 'EnvironmentSatisfaction', 'JobInvolvement', 'JobSatisfaction', 'PercentSalaryHike', 'PerformanceRating', 'RelationshipSatisfaction', 'TrainingTimesLastYear', 'WorkLifeBalance']


# Create X_df using your selected columns
X_df = attrition_df[X_list]

# Show the data types for X_df
X_df.dtypes


Age                         int64
DistanceFromHome            int64
EnvironmentSatisfaction     int64
JobInvolvement              int64
JobSatisfaction             int64
PercentSalaryHike           int64
PerformanceRating           int64
RelationshipSatisfaction    int64
TrainingTimesLastYear       int64
WorkLifeBalance             int64
dtype: object

In [5]:
X_df

Age  DistanceFromHome  EnvironmentSatisfaction  JobInvolvement  \
0      41                 1                        2               3   
1      49                 8                        3               2   
2      37                 2                        4               2   
3      33                 3                        4               3   
4      27                 2                        1               3   
...   ...               ...                      ...             ...   
1465   36                23                        3               4   
1466   39                 6                        4               2   
1467   27                 4                        2               4   
1468   49                 2                        4               2   
1469   34                 8                        2               4   

      JobSatisfaction  PercentSalaryHike  PerformanceRating  \
0                   4                 11                  3   
1                   2                 23                  4   
2                   3                 15                  3   
3                   3                 11                  3   
4                   2                 12                  3   
...               ...                ...                ...   
1465                4                 17                  3   
1466                1                 15                  3   
1467                2                 20                  4   
1468                2                 14                  3   
1469                3                 12                  3   

      RelationshipSatisfaction  TrainingTimesLastYear  WorkLifeBalance  
0                            1                      0                1  
1                            4                      3                3  
2                            2                      3                3  
3                            3                      3                3  
4                            4                      3                3  
...                        ...                    ...              ...  
1465                         3                      3                3  
1466                         1                      5                3  
1467                         2                      0                3  
1468                         4                      3                2  
1469                         1                      3                4  

[1470 rows x 10 columns]

In [6]:
X=X_df
y=y_df

In [7]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=78)



In [8]:
X_train


Age  DistanceFromHome  EnvironmentSatisfaction  JobInvolvement  \
1304   40                12                        2               3   
1125   29                 2                        1               3   
139    30                 9                        3               3   
499    33                 8                        3               3   
832    37                25                        3               2   
...   ...               ...                      ...             ...   
310    31                 2                        1               2   
105    59                 2                        3               2   
40     35                 4                        3               3   
470    24                24                        4               3   
709    31                 9                        3               2   

      JobSatisfaction  PercentSalaryHike  PerformanceRating  \
1304                1                 12                  3   
1125                4                 19                  3   
139                 4                 19                  3   
499                 3                 12                  3   
832                 4                 13                  3   
...               ...                ...                ...   
310                 1                 12                  3   
105                 4                 21                  4   
40                  4                 12                  3   
470                 4                 13                  3   
709                 1                 22                  4   

      RelationshipSatisfaction  TrainingTimesLastYear  WorkLifeBalance  
1304                         2                      3                3  
1125                         4                      0                4  
139                          4                      2                1  
499                          4                      3                3  
832                          3                      2                3  
...                        ...                    ...              ...  
310                          4                      1                3  
105                          4                      3                3  
40                           3                      3                3  
470                          3                      3                3  
709                          1                      0                3  

[1176 rows x 10 columns]

In [9]:
X_test

Age  DistanceFromHome  EnvironmentSatisfaction  JobInvolvement  \
471    38                10                        3               3   
1090   29                 8                        3               1   
441    42                 5                        2               3   
979    34                29                        2               3   
1085   31                 3                        4               3   
...   ...               ...                      ...             ...   
751    50                 1                        4               3   
728    54                17                        3               3   
63     59                25                        1               3   
1008   54                 1                        4               3   
524    34                 9                        4               2   

      JobSatisfaction  PercentSalaryHike  PerformanceRating  \
471                 3                 19                  3   
1090                1                 13                  3   
441                 3                 17                  3   
979                 3                 13                  3   
1085                3                 12                  3   
...               ...                ...                ...   
751                 3                 12                  3   
728                 3                 11                  3   
63                  1                 11                  3   
1008                4                 19                  3   
524                 2                 14                  3   

      RelationshipSatisfaction  TrainingTimesLastYear  WorkLifeBalance  
471                          3                      4                3  
1090                         3                      3                3  
441                          4                      4                3  
979                          1                      1                3  
1085                         1                      2                1  
...                        ...                    ...              ...  
751                          1                      3                3  
728                          3                      2                3  
63                           4                      3                2  
1008                         4                      3                2  
524                          2                      3                4  

[294 rows x 10 columns]

In [10]:
y_train

Attrition              Department
1304        No  Research & Development
1125        No  Research & Development
139         No         Human Resources
499         No                   Sales
832         No  Research & Development
...        ...                     ...
310         No         Human Resources
105         No         Human Resources
40          No  Research & Development
470         No                   Sales
709        Yes  Research & Development

[1176 rows x 2 columns]

In [11]:
y_test

Attrition              Department
471         No  Research & Development
1090        No  Research & Development
441         No  Research & Development
979         No  Research & Development
1085       Yes  Research & Development
...        ...                     ...
751         No                   Sales
728         No  Research & Development
63          No                   Sales
1008        No  Research & Development
524         No  Research & Development

[294 rows x 2 columns]

In [12]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary


***Note that all selected columns have values, Int64, so no further action is necessary to convert to numberical data types.***

In [13]:
# Create a StandardScaler
scaler = StandardScaler()

# Fit the StandardScaler to the training data
X_scaler = scaler.fit(X_train)

# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)



In [14]:
X_train_scaled

array([[ 3.57156995e-01,  3.77850497e-01, -6.42037467e-01, ...,
        -6.59449578e-01,  1.53905723e-01,  3.39490953e-01],
       [-8.64144324e-01, -8.79868677e-01, -1.55834337e+00, ...,
         1.18921003e+00, -2.18652268e+00,  1.71618530e+00],
       [-7.53116932e-01,  5.34744547e-04,  2.74268433e-01, ...,
         1.18921003e+00, -6.26237078e-01, -2.41389774e+00],
       ...,
       [-1.97979968e-01, -6.28324842e-01,  2.74268433e-01, ...,
         2.64880224e-01,  1.53905723e-01,  3.39490953e-01],
       [-1.41928129e+00,  1.88711351e+00,  1.19057433e+00, ...,
         2.64880224e-01,  1.53905723e-01,  3.39490953e-01],
       [-6.42089539e-01,  5.34744547e-04,  2.74268433e-01, ...,
        -1.58377938e+00, -2.18652268e+00,  3.39490953e-01]])

In [15]:
# Create a OneHotEncoder for the Department column
from sklearn.preprocessing import OneHotEncoder

department_encoder = OneHotEncoder(sparse_output=False)

# Fit the encoder to the training data
# Create two new variables by applying the encoder
# to the training and testing data
department_train=department_encoder.fit_transform(y_train[['Department']])
department_test=department_encoder.fit_transform(y_test[['Department']])

department_train



array([[0., 1., 0.],
       [0., 1., 0.],
       [1., 0., 0.],
       ...,
       [0., 1., 0.],
       [0., 0., 1.],
       [0., 1., 0.]])

In [16]:
# Convert the Department OneHotEncoded Arrays back to data frames
department_train_df = pd.DataFrame(department_train,index=y_train.index,columns=department_encoder.get_feature_names_out(['Department']))

department_test_df = pd.DataFrame(department_test,index=y_test.index,columns=department_encoder.get_feature_names_out(['Department']))

department_train_df


Department_Human Resources  Department_Research & Development  \
1304                         0.0                                1.0   
1125                         0.0                                1.0   
139                          1.0                                0.0   
499                          0.0                                0.0   
832                          0.0                                1.0   
...                          ...                                ...   
310                          1.0                                0.0   
105                          1.0                                0.0   
40                           0.0                                1.0   
470                          0.0                                0.0   
709                          0.0                                1.0   

      Department_Sales  
1304               0.0  
1125               0.0  
139                0.0  
499                1.0  
832                0.0  
...                ...  
310                0.0  
105                0.0  
40                 0.0  
470                1.0  
709                0.0  

[1176 rows x 3 columns]

In [17]:
department_test_df

Department_Human Resources  Department_Research & Development  \
471                          0.0                                1.0   
1090                         0.0                                1.0   
441                          0.0                                1.0   
979                          0.0                                1.0   
1085                         0.0                                1.0   
...                          ...                                ...   
751                          0.0                                0.0   
728                          0.0                                1.0   
63                           0.0                                0.0   
1008                         0.0                                1.0   
524                          0.0                                1.0   

      Department_Sales  
471                0.0  
1090               0.0  
441                0.0  
979                0.0  
1085               0.0  
...                ...  
751                1.0  
728                0.0  
63                 1.0  
1008               0.0  
524                0.0  

[294 rows x 3 columns]

In [18]:
# Create a OneHotEncoder for the Attrition column
attrition_encoder = OneHotEncoder(sparse_output=False)

# Fit the encoder to the training data
# Create two new variables by applying the encoder
# to the training and testing data
attrition_train=attrition_encoder.fit_transform(y_train[['Attrition']])
attrition_test=attrition_encoder.fit_transform(y_test[['Attrition']])

attrition_train


array([[1., 0.],
       [1., 0.],
       [1., 0.],
       ...,
       [1., 0.],
       [1., 0.],
       [0., 1.]])

## Create, Compile, and Train the Model

In [19]:
# Convert the Attrition OneHotEncoded Arrays back to data frames
attrition_train_df = pd.DataFrame(attrition_train,index=y_train.index,columns=attrition_encoder.get_feature_names_out(['Attrition']))

attrition_test_df = pd.DataFrame(attrition_test,index=y_test.index,columns=attrition_encoder.get_feature_names_out(['Attrition']))

attrition_train_df

Attrition_No  Attrition_Yes
1304           1.0            0.0
1125           1.0            0.0
139            1.0            0.0
499            1.0            0.0
832            1.0            0.0
...            ...            ...
310            1.0            0.0
105            1.0            0.0
40             1.0            0.0
470            1.0            0.0
709            0.0            1.0

[1176 rows x 2 columns]

In [20]:
attrition_test_df

Attrition_No  Attrition_Yes
471            1.0            0.0
1090           1.0            0.0
441            1.0            0.0
979            1.0            0.0
1085           0.0            1.0
...            ...            ...
751            1.0            0.0
728            1.0            0.0
63             1.0            0.0
1008           1.0            0.0
524            1.0            0.0

[294 rows x 2 columns]

In [21]:
# Concatenate the one-hot encoded department and attrition data with the y_train and y_test
y_train_encoded=pd.concat([y_train,department_train_df,attrition_train_df],axis=1)
y_test_encoded=pd.concat([y_test,department_test_df,attrition_test_df],axis=1)

y_train_encoded


Attrition              Department  Department_Human Resources  \
1304        No  Research & Development                         0.0   
1125        No  Research & Development                         0.0   
139         No         Human Resources                         1.0   
499         No                   Sales                         0.0   
832         No  Research & Development                         0.0   
...        ...                     ...                         ...   
310         No         Human Resources                         1.0   
105         No         Human Resources                         1.0   
40          No  Research & Development                         0.0   
470         No                   Sales                         0.0   
709        Yes  Research & Development                         0.0   

      Department_Research & Development  Department_Sales  Attrition_No  \
1304                                1.0               0.0           1.0   
1125                                1.0               0.0           1.0   
139                                 0.0               0.0           1.0   
499                                 0.0               1.0           1.0   
832                                 1.0               0.0           1.0   
...                                 ...               ...           ...   
310                                 0.0               0.0           1.0   
105                                 0.0               0.0           1.0   
40                                  1.0               0.0           1.0   
470                                 0.0               1.0           1.0   
709                                 1.0               0.0           0.0   

      Attrition_Yes  
1304            0.0  
1125            0.0  
139             0.0  
499             0.0  
832             0.0  
...             ...  
310             0.0  
105             0.0  
40              0.0  
470             0.0  
709             1.0  

[1176 rows x 7 columns]

In [22]:
# Find the number of columns in the X training data
num_columns = X_train.shape[1]
print(num_columns)

# Create the input layer
input_layer = layers.Input(shape=(num_columns,))

# Create at least two shared layers
shared_layer1 = layers.Dense(units=64, activation='relu', name = 'shared_layer1')(input_layer)
shared_layer2 = layers.Dense(units=128, activation='relu', name = 'shared_layer2')(shared_layer1)

10


In [23]:
# Create a branch for Department
# with a hidden layer and an output layer

# Create the hidden layer
department_hidden = layers.Dense(units=32, activation='softmax', name='department_hidden')(shared_layer2)

# Create the output layer
department_output = layers.Dense(units=3, activation='softmax', name='department_output')(department_hidden)



In [24]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer
attrition_hidden = layers.Dense(units=32, activation='sigmoid', name='attrition_hidden')(shared_layer2)

# Create the output layer
attrition_output = layers.Dense(units=2, activation='sigmoid', name='attrition_output')(attrition_hidden)


In [25]:
# Create the model
model = Model(inputs=input_layer, outputs=[department_output, attrition_output])

# Compile the model
model.compile(optimizer='adam', loss={'department_output': 'categorical_crossentropy',
                    'attrition_output': 'binary_crossentropy'},
              metrics={'department_output': 'accuracy',
                        'attrition_output': 'accuracy'})

# Summarize the model
model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)              ┃ Output Shape           ┃        Param # ┃ Connected to           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)  │ (None, 10)             │              0 │ -                      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ shared_layer1 (Dense)     │ (None, 64)             │            704 │ input_layer[0][0]      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ shared_layer2 (Dense)     │ (None, 128)            │          8,320 │ shared_layer1[0][0]    │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ department_hidden (Dense) │ (None, 32)             │          4,128 │ shared_layer2[0][0]    │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ attrition_hidden (Dense)  │ (None, 32)             │          4,128 │ shared_layer2[0][0]    │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ department_output (Dense) │ (None, 3)              │             99 │ department_hidden[0][… │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ attrition_output (Dense)  │ (None, 2)              │             66 │ attrition_hidden[0][0] │
└───────────────────────────┴────────────────────────┴────────────────┴────────────────────────┘

 Total params: 17,445 (68.14 KB)

 Trainable params: 17,445 (68.14 KB)

 Non-trainable params: 0 (0.00 B)

In [26]:
# Train the model
model.fit(X_train_scaled,
          {'department_output': department_train,
           'attrition_output': attrition_train},
          epochs=100,
          batch_size=32,
          validation_split=0.2)


Epoch 1/100
30/30 ━━━━━━━━━━━━━━━━━━━━ 3s 15ms/step - attrition_output_accuracy: 0.4585 - department_output_accuracy: 0.4296 - loss: 1.7936 - val_attrition_output_accuracy: 0.8263 - val_department_output_accuracy: 0.6483 - val_loss: 1.4809
Epoch 2/100
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - attrition_output_accuracy: 0.8300 - department_output_accuracy: 0.7007 - loss: 1.4287 - val_attrition_output_accuracy: 0.8263 - val_department_output_accuracy: 0.6483 - val_loss: 1.3758
Epoch 3/100
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - attrition_output_accuracy: 0.8289 - department_output_accuracy: 0.6554 - loss: 1.3601 - val_attrition_output_accuracy: 0.8263 - val_department_output_accuracy: 0.6483 - val_loss: 1.3304
Epoch 4/100
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - attrition_output_accuracy: 0.8294 - department_output_accuracy: 0.6630 - loss: 1.3087 - val_attrition_output_accuracy: 0.8263 - val_department_output_accuracy: 0.6483 - val_loss: 1.3033
Epoch 5/100
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 5

In [27]:
# Evaluate the model with the testing data
test_results = model.evaluate(X_test_scaled, {'attrition_output': attrition_test, 'department_output': department_test})
test_results

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - attrition_output_accuracy: 0.7839 - department_output_accuracy: 0.5350 - loss: 1.9593 


[1.8880847692489624, 0.7891156673431396, 0.5646258592605591]

In [28]:
# Print the accuracy for both department and attrition
print(f"Attrition Predictions Accuracy: {test_results[1]}")
print(f"Department Predictions Accuracy: {test_results[2]}")

Attrition Predictions Accuracy: 0.7891156673431396
Department Predictions Accuracy: 0.5646258592605591


In [29]:
value_counts_dept = y_train['Department'].value_counts()
print(value_counts_dept)

Department
Research & Development    780
Sales                     349
Human Resources            47
Name: count, dtype: int64


In [30]:
value_counts_attr = y_train['Attrition'].value_counts()
print(value_counts_attr)

Attrition
No     981
Yes    195
Name: count, dtype: int64


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1. Accuracy may not be the best metric to use on this data.  This is specifically due to the fact that Accuracy works best when classes are balanced.  There is a risk that the accuracy metric may be skewed in an imbalanced data set simply by the tendency to predict the majority class.  In the predictions of this challenge, the attrition target is significantly skewed toward a "no" prediction as the majority class, making up 83% of the training set.  Similarly, Research & Development is highly skewed in the department target, making up 66% of the training set.
2. I chose the activation function of Sigmoid for the 'Attrition' target in the output layers.  And, I chose the Softmax activation function for the output layers of the 'Department' target.  In general, Sigmoid is recommended for binary data, which is true for the Attrition target, with Yes or No (0 or 1 after One-Hot Encoding) as the only 2 possible results (binary).  In contrast, Softmax is generally recommended for multi-class data sets.  The 'Department' results are multi-class, with 3 possible classifications (Research & Development, Sales, & Human Resources).  
3. A few ways that the model may be improved include the following;      1. The feature set can be modified experiment with the 10 selected features, or to use a different set of features to predict 'Attrition" than used to predict the most suitable 'Department.'    2. Apply RandomOversampling or RandomUndersampling to balance the data.  This may help to address any impact of the majority class skewing the accuracy metric in making predictions,    3. Experiment with the batch-sizes or number of epochs to improve the model,    4. Add more layers or neurons to the model to ensure more complex aspects of the data are addressed. 